src/chain/shared_state.rs

```diff
@@ -0,0 +1,174 @@
+pub struct SharedState {
+    /// The new self section info, that doesn't necessarily have a full set of signatures yet.
+    pub new_info: SectionInfo,
+    /// The latest few fully signed infos of our own sections, each with signatures by the previous
+    /// one. This is included in every message we relay.
+    /// This is not a `BTreeSet` just now as it is ordered according to the sequence of pushes into
+    /// it.
+    pub our_infos: Vec<(SectionInfo, ProofSet)>,
+    /// Any change (split or merge) to the section that is currently in progress.
+    pub change: SectionChange,
+    // The accumulated `SectionInfo`(self or sibling) and proofs during a split pfx change.
+    pub split_cache: Option<(SectionInfo, ProofSet)>,
+    /// The set of section info hashes that are currently merging.
+    pub merging: BTreeSet<Digest256>,
+}
+
+impl SharedState {
+    pub fn new(section_info: SectionInfo) -> Self {
+        Self {
+            new_info: section_info.clone(),
+            our_infos: vec![(section_info, Default::default())],
+            change: SectionChange::None,
+            split_cache: None,
+            merging: Default::default(),
+        }
+    }
+
+    pub fn our_infos(
+        &self,
+    ) -> impl Iterator<Item = &SectionInfo> + ExactSizeIterator + DoubleEndedIterator {
+        self.our_infos.iter().map(|(si, _)| si)
+    }
+
+    pub fn opt_our_info(&self) -> Option<&SectionInfo> {
+        self.our_infos.last().map(|(si, _)| si)
+    }
+
+    /// Returns our own current section info.
+    pub fn our_info(&self) -> &SectionInfo {
+        // TODO: Replace `our_infos` with a new `NonemptyVec` type that statically guarantees that
+        // it's never empty.
+        unwrap!(self.opt_our_info())
+    }
+
+    pub fn our_prefix(&self) -> &Prefix<XorName> {
+        self.opt_our_info()
+            .map_or(&DEFAULT_PREFIX, SectionInfo::prefix)
+    }
+
+    pub fn our_version(&self) -> u64 {
+        self.opt_our_info().map_or(0, |si| *si.version())
+    }
+
+    /// Returns our section info with the given hash, if it exists.
+    pub fn our_info_by_hash(&self, hash: &Digest256) -> Option<&SectionInfo> {
+        self.our_infos
+            .iter()
+            .find(|(sec_info, _)| sec_info.hash() == hash)
+            .map(|(sec_info, _)| sec_info)
+    }
+
+    /// Returns the section info matching our own name with the given version number.
+    pub fn our_info_by_version(&self, version: u64) -> Option<&SectionInfo> {
+        // TODO: Binary search? Reverse order? Benchmark which one is fastest in practice.
+        self.our_infos
+            .iter()
+            .find(|(sec_info, _)| *sec_info.version() == version)
+            .map(|(sec_info, _)| sec_info)
+    }
+
+    /// Remove our infos that are older than the given version.
+    pub(super) fn clean_our_infos(&mut self, oldest_version: u64) {
+        // TODO since the list is sorted according to the version, choose something more efficient
```

Comment:
Will this clash with PR 1635. (We probably should rebase 1647 on top of it).
Otherwise I'd probably say the same.

Reply:
I'll address it when #1635 is merged.